string_test.go

```diff
@@ -103,3 +103,31 @@ func TestString_EndsWith(t *testing.T) {
 	assert.False(t, s.Parse("TEST").IsValid())
 	assert.False(t, s.Parse("3tes3t").IsValid())
 }
+
+func TestString_Optional(t *testing.T) {
```

Comment:
I think it's better if this test case is split out, I think it's trying to do too much.

since it's named `TestString_Optional`, we should just be testing optional strings, we're retesting a lot of things that have been tested in other cases such as `TestString_Includes` where we test whether a string includes a value. If we treat this test case as more pure we could end up with an implementation like

```go
func TestString_Optional(t *testing.T) {
	s := schema.String().Optional()

	var nullValue *string = nil

	assert.True(t, s.Parse(nullValue).IsValid())
}
```

which is much smaller and to the point, hopefully making it more readable what the intention is, to allow `nil` value when specifying the optional field. You could retest non-optional in here as well by defining another schema without `.Optional` just to be 100% sure but we test this case in the `TestString_Type` function, where to assert that `nil` should result in a `false` condition.

There are some test cases that are returning false positives as well, 

```go
v2 := "X_test_X"
assert.True(t, s.Parse(&v2).IsValid()) // Pointer to valid string
```

The issue with the above is that it's returning a false positive due to reusing the schema above, which is defined as `Optional` this means it's not actually testing whether the `Parse` function is capable of taking the value from a pointer, and performing validation on that, since we've said it can be `nil`. If we were to write a test like the below, it would error

```go
func TestString_Pointer(t *testing.T) {
	s := schema.String()

	value := "test"

	assert.True(t, s.Parse(&value).IsValid())
}
```

This test would actually fail because the `Parse` function would return `false` here, when this is a perfectly valid use-case.

Reply:
Yeah i see what you mean, I think the complication within this test comes from trying to test two types of functionality supporting both pointers and then nil values in the same test as you referenced here https://github.com/Jamess-Lucass/validator-go/pull/18/files#r1825266146

Reply:
I'd need to support pointer values for everything whether its optional or not